### PR TITLE
 	storage: Be more careful with mdraid disks

### PR DESCRIPTION
**Pull request by mvollmer**

### CODE REVIEW

**Comment by mvollmer** [comment]
@andreasn, please check the texts.

**Comment by mvollmer** [comment]
Needs tests, tests, tests, ...

**Comment by mvollmer** [comment]
> Needs tests, tests, tests, ...

Fixed tests and added some more.

**Comment by mvollmer** [comment]
@andreasn, the texts are as follows:

While the array is stopped: "The MDRAID device must be running in order to remove disks." and "The MDRAID device must be running in order to add spare disks."  We need a running array to determine whether removing a disk is safe, and there is no way to add a disk to a stopped array.

For a disk that is "in sync" (holding data): "A spare disk needs to be added first before this disk can be removed."  The idea is that the spare will immediately take over.

While the array is recovering: "Disks can not be removed while the device is recovering."  I think it is best not to allow this.  Just a feeling, though.

For the last disk: "The last disk of a MDRAID device cannot be removed."

**Comment by mvollmer** [comment]
> While the array is recovering: "Disks can not be removed while the device is recovering." I think it is best not to allow this. Just a feeling, though.

I have think more about this case...

**Comment by mvollmer** [comment]
> While the array is recovering: "Disks can not be removed while the device is recovering." I think it is best not to allow this. Just a feeling, though.

Okay, failed disks and spares can be removed while an array is recovering, anything else either doesn't work or will destroy the array.  I have updated the condition in the code, and changed the text to "This disk can not be removed while the device is recovering."

File: pkg/storaged/details.js [hunk 1]
```diff
@@ -63,7 +63,7 @@
             mdraid_add_disk: function mdraid_add_disk(path) {
                 var mdraid = client.mdraids[path];
 
-                dialog.open({ Title: _("Add Disks"),
+                dialog.open({ Title: _("Add Spare Disks"),
```
Comment (dperpeet) [inline pkg/storaged/details.js]: This title can be misleading. If a RAID is degraded (missing disks), the disk I want to add might not be a spare at all, but essential.

File: pkg/storaged/details.js [hunk 2]
```diff
@@ -418,13 +428,26 @@
...
                 }
 
+                var is_in_sync = (active_state && active_state[2].indexOf("in_sync") >= 0);
+                var is_recovering = (active_state && active_state[2].indexOf("spare") >= 0 && active_state[1] >= 0);
+
+                var excuse = false;
+                if (!running)
+                    excuse = _("The MDRAID device must be running in order to remove disks.");
+                else if ((is_in_sync && n_recovering > 0) || is_recovering)
+                    excuse = _("This disk can not be removed while the device is recovering.");
```
Comment (dperpeet) [inline pkg/storaged/details.js]: `cannot` or `can't`

File: pkg/storaged/details.js (hunk 1 above)
Comment (dperpeet) [inline pkg/storaged/details.js]: @garrett What do you think?

**Comment by andreasn** [comment]
In the panel we call it "[name] RAID Device", but in the tooltip of this PR it's called "MDRAID device". (MD as in Multiple Device/Disk I assume)
Probably best to keep these consistent. I would go for "RAID Device", as that's what I've seen most documentation refer to it as.

**Comment by andreasn** [comment]
Apart from that, this looks good I think.

File: pkg/storaged/details.js (hunk 1 above)
Comment (mvollmer) [inline pkg/storaged/details.js]: Yeah, hmm.  The distinction I am trying to make is between adding a disk in order to make the RAID device larger, and adding a disk in order to add more redundancy.

**Comment by mvollmer** [comment]
> Probably best to keep these consistent. I would go for "RAID Device", as that's what I've seen most documentation refer to it as.

Yes, very good!

File: pkg/storaged/details.js (hunk 2 above)
Comment (mvollmer) [inline pkg/storaged/details.js]: cannot!  Thanks.

File: pkg/storaged/details.js (hunk 1 above)
Comment (mvollmer) [inline pkg/storaged/details.js]: I am okay with not changing the title at all, and leaving it as "Add Disks".

**Comment by mvollmer** [comment]
(Yes, we got the connection.id failure with a backtrace!)

File: pkg/storaged/details.js (hunk 1 above)
Comment (andreasn) [inline pkg/storaged/details.js]: Add Disks sounds good to me.

Comment (mvollmer) [inline pkg/storaged/details.js]: Okay, removed that change.

**Pull request closed without merging**